moquette-0.17/broker/src/main/java/io/moquette/broker/unsafequeues/QueuePool.java

```diff
@@ -0,0 +1,490 @@
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            QueueName queueName = (QueueName) o;
+            return Objects.equals(name, queueName.name);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(name);
+        }
+
+        @Override
+        public String toString() {
+            return "QueueName{name='" + name + '\'' + '}';
+        }
+    }
+
+    private final SegmentAllocator allocator;
+    private final Path dataPath;
+    private final int segmentSize;
+    private final ConcurrentMap<QueueName, LinkedList<SegmentRef>> queueSegments = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueueName, Queue> queues = new ConcurrentHashMap<>();
+    private final ConcurrentSkipListSet<SegmentRef> recycledSegments = new ConcurrentSkipListSet<>();
+    private final ReentrantLock segmentsAllocationLock = new ReentrantLock();
+
+    private QueuePool(SegmentAllocator allocator, Path dataPath, int segmentSize) {
+        this.allocator = allocator;
+        this.dataPath = dataPath;
+        this.segmentSize = segmentSize;
+        this.callback = new SegmentAllocationCallback(this);
+    }
+
+    private static class SegmentAllocationCallback implements PagedFilesAllocator.AllocationListener {
+
+        private final QueuePool queuePool;
+
+        private SegmentAllocationCallback(QueuePool queuePool) {
+            this.queuePool = queuePool;
+        }
+
+        @Override
+        public void segmentedCreated(String name, Segment segment) {
+            queuePool.segmentedCreated(name, segment);
+        }
+    }
+
+    private void segmentedCreated(String name, Segment segment) {
+        LOG.debug("Registering new segment {} for queue {}", segment, name);
+        final QueueName queueName = new QueueName(name);
+        List<SegmentRef> segmentRefs = this.queueSegments.computeIfAbsent(queueName, k -> new LinkedList<>());
+
+        // adds in head
+        segmentRefs.add(0, new SegmentRef(segment));
+
+        LOG.debug("queueSegments for queue {} after insertion {}", queueName, segmentRefs);
+    }
+
+    public static QueuePool loadQueues(Path dataPath, int pageSize, int segmentSize) throws QueueException {
+        // read in checkpoint.properties
+        final Properties checkpointProps = createOrLoadCheckpointFile(dataPath);
+
+        // load last references to segment and instantiate the allocator
+        final int lastPage = Integer.parseInt(checkpointProps.getProperty("segments.last_page", "0"));
+        final int lastSegment = Integer.parseInt(checkpointProps.getProperty("segments.last_segment", "0"));
+
+        final PagedFilesAllocator allocator = new PagedFilesAllocator(dataPath, pageSize, segmentSize, lastPage, lastSegment);
+
+        final QueuePool queuePool = new QueuePool(allocator, dataPath, segmentSize);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource: *queuePool*. The resource is referenced at lines 145, 146, 147, 149, 150. The resource is returned at line 151. There are other execution paths that don&#39;t close the resource or return it, for example, when QueuePool.loadQueueDefinitions throws an exception. To prevent this resource leak, close *queuePool* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).